### PR TITLE
Document card search

### PR DESCRIPTION
**Pull request by lukemelia**

### CODE REVIEW

**Comment by lukemelia** [comment]
I'd love some initial input regarding the TODOs in this, and what else I should add.

File: docs/search.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+* `type`: allow specification of a particular card type. Also includes descendant cards in this card's adoption chain.
+* `eq`: allows specifying a card with a specific field value
+* `range`: allows specifying a card with a field value that is gt, gte, lt, or lte a specific value
+* `any`: allows an "OR" union of other filters
+* `every`: allows an "AND" union of other filters
+* `not`: allow negating another filter
+
+* `on`: may be used with `eq` or `range` to limit results to a particular card type.
+
+Nested fields may be specified using a dot (`.`) notation.
```
Comment (habdelra) [inline docs/search.md:24]: might be worth mentioning that our ability to filter containsMany and linksToMany fields is quite limited. Say for instance we have a card `Company`  with a `department` field that is a linksToMany field. There would be no easy way to express a filter that returns all the companies that have more than 5 departments. So the idea is that better support for plural field predicate filtering is desired.

File: docs/search.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+The Query object may have a `filter` property, which controls which cards are included in the result set.
+
+* `type`: allow specification of a particular card type. Also includes descendant cards in this card's adoption chain.
+* `eq`: allows specifying a card with a specific field value
+* `range`: allows specifying a card with a field value that is gt, gte, lt, or lte a specific value
+* `any`: allows an "OR" union of other filters
+* `every`: allows an "AND" union of other filters
+* `not`: allow negating another filter
+
+* `on`: may be used with `eq` or `range` to limit results to a particular card type.
```
Comment (habdelra) [inline docs/search.md]: `on` is a lot more meaningful that what is being decribed here, i think we need to say a stronger thing. `on` provides the card type context for a field. This is how we can disambiguate between a `Company.name` field and a `Country.name`. the `name` field field doesn't live in isolation. simply providing a predicate to filter by `name` field isnt enough. You need to say which card's `name` field you want to filter by--this is the function of `on`. It imbues the card definition context to fields specified in the filter(s).

File: docs/search.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+      {
+        by: 'author.lastName',
+        on: { module: `https://my.realm/book`, name: 'Book' },
+      },
+    ],
+    filter: { type: { module: `https://my.realm/book`, name: 'Book' } },
+  });
+```
+
+### Pagination -- TODO: is this used yet????
```
Comment (habdelra) [inline docs/search.md]: yeah this is still TBD. I'm guessing we wont really tackle this until we have a DB based index

File: docs/search.md [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+### queryString -- TODO: is this used????
```
Comment (habdelra) [inline docs/search.md]: We do have support for a serialized form of our filters  as a query string, that is how our card service is able to perform queries: https://github.com/cardstack/boxel/blob/f76dbedc9e828ff2b5ced0f57406e0fd4ae49b28/packages/host/app/services/card-service.ts#L151 Is that what you mean here?

**Comment by habdelra** [comment]
Also probably worth mentioning, currently the search index doesn't cache the linked fields. those are, for the time being, considered volatile and will be loaded at the time the request is made. In the future we anticipate providing caching for non-volatile linked fields.

File: docs/search.md (hunk 1 above)
Comment (lukemelia) [reply docs/search.md:24]: 👍

File: docs/search.md (hunk 2 above)
Comment (lukemelia) [reply docs/search.md]: 👍

File: docs/search.md (hunk 3 above)
Comment (lukemelia) [reply docs/search.md]: 👍

File: docs/search.md (hunk 4 above)
Comment (lukemelia) [reply docs/search.md]: I was thinking of this: https://github.com/cardstack/boxel/blob/master/packages/runtime-common/query.ts#L11. I can't find any evidence of it being used.

**Comment by lukemelia** [comment]
> Also probably worth mentioning, currently the search index doesn't cache the linked fields. those are, for the time being, considered volatile and will be loaded at the time the request is made. In the future we anticipate providing caching for non-volatile linked fields.

Does this mean that you can't used nested field syntax to search them, or that the searches will be inefficient, or both?

**Comment by habdelra** [comment]
> Does this mean that you can't used nested field syntax to search them, or that the searches will be inefficient, or both?

when we build up the search doc for a card with nested fields we will accurately reflect that as well as the dependent cards. When the dependent cards are updated, we invalidate the all the cards that consume them and rebuild their search docs--so searches will be correct. We don't, however, cache JSON API documents for the cards in the index--rather we cache just the JSON API resource (so sans the `included` part of the JSON API document). so when serving the card content from the realm, we reconstruct the `included` part of the JSON API document each time. So, yeah, it's inefficient.

File: docs/search.md (hunk 4 above)
Comment (habdelra) [reply docs/search.md]: hmmm, I'm not sure what that is actually. we copy pasted this module from the previous incarnation of the hub. perhaps that had a meaning there (I forget)...

File: docs/search.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+### Sort Order
+
+The Query object may have a `sort` property, which controls the order in which results are returned.
+
+A `sort` is an array property value where each item is an object that consists of `on`, `by`, and optionally `direction`. The array is the order in which the sort is applied.
+
+* `on`: The card type being sorted
+* `by`: The field name to be sorted on. Nested fields may be supported using a dot (`.`) notation.
+* `direction`: `asc` of `desc`, default is `asc`
```
Comment (backspace) [inline docs/search.md]:
```suggestion
* `direction`: `asc` or `desc`, default is `asc`
```

**Review by backspace — APPROVED** [review]

**Review by FadhlanR — APPROVED** [review]

**Review by jurgenwerk — APPROVED** [review]

**Pull request merged**